pypsa/inverter_functions.py

```diff
@@ -0,0 +1,512 @@
+    """
+    # parameters needed
+    set_p1 = c_attrs['set_p1']
+    set_p2 = c_attrs['set_p2']
+    s_nom = c_attrs['s_nom']
+    power_factor_min = c_attrs['power_factor_min']
+    p_set_per_p_ref = (abs(p_input.loc[now, c_attrs.index]) / c_attrs['p_ref'])*100
+
+    # choice of power_factor according to controller inputs and its droop curve
+    power_factor = np.select([(p_set_per_p_ref < set_p1), (
+        p_set_per_p_ref >= set_p1) & (p_set_per_p_ref <= set_p2), (
+            p_set_per_p_ref > set_p2)], [1, (1 - ((1 - power_factor_min) / (
+             set_p2 - set_p1) * (p_set_per_p_ref - set_p1))), power_factor_min])
+
+    # find q_set and avoid -0 apperance as the output when power_factor = 1
+    q_set = np.where(power_factor == 1, 0, -p_input.loc[
+            now, c_attrs.index].mul(np.tan((np.arccos(
+                          power_factor, dtype=np.float64)), dtype=np.float64)))
+
+    S = np.sqrt((p_input.loc[now, c_attrs.index])**2 + q_set**2)
+    assert ((S < s_nom).any().any()), (
+        "The resulting reactive power (q)  while using 'cosphi'_p control  "
+        "with the chosen attr 'power_factor_min' in '%s' component results a  "
+        "complex power (S = sqrt(p**2 + q**2))) which is greater than 's_nom') "
+        "of the inverter, please choose the right power_factor_min value"
+        % (c_list_name))
+
+    # set the the resulting power factor output in n.components(_t).power_factor
+    if time_varying_p_set:
+        df_t.power_factor.loc[now, c_attrs.index] = power_factor
+    else:
+        df.loc[c_attrs.index, 'power_factor'] = power_factor
+
+    return q_set
+
+
+def find_allowable_q(p, power_factor, s_nom):
+
+    Q_max = p.mul(np.tan((np.arccos(power_factor, dtype=np.float64)),
+                         dtype=np.float64))
+    # find inverter q capacity according to power factor provided
+    Q_inv_cap = s_nom*np.sin(np.arccos(power_factor, dtype=np.float64),
+                             dtype=np.float64)
+    # find max allowable q that is possible based on s_nom
+    Q_allowable = np.where(Q_max <= Q_inv_cap, Q_max, Q_inv_cap)
+
+    return Q_inv_cap, Q_allowable, Q_max
+
+
+def adjust_p_set(s_nom, q_set, p, c_list_name, control_type):
+
+    adjusted_p_set = np.sqrt((s_nom**2 - q_set**2),  dtype=np.float64)
+    new_p_set = np.where(abs(p) <= abs(adjusted_p_set), p, adjusted_p_set)
+
+    log_info = np.where(
+            control_type == 'fixed_cosphi', '"fixed_cosphi" control is adjusted',
+            ' "q_v" control might be adjusted, if needed')
+
+    logger.info(" Some p_set in '%s' component with %s due to reactive power "
+                "compensation priority. ", c_list_name, log_info)
+
+    return new_p_set
+
+
+def q_v(c_list_name, now, n_trials_max, n_trials, p_input, v_pu_bus, c_attrs):
+    """
+    Reactive power as a function of voltage Q(U).
+    reference : https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=6096349
```

Comment:
Same as above. Provide a brief description of the control strategy here, but feel free to keep the paper as additional reference. Maybe provide the doi link instead.